### PR TITLE
Added_Schemas

### PR DESCRIPTION
**Pull request by lakshay42jain**

### CODE REVIEW

File: test/fixtures/task_categories.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
```
Comment (gagan93jtg) [inline test/fixtures/task_categories.yml]: don't use fixtures. Use factories (later when u start working on test cases)

File: test/models/task_comment_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class TaskCommentTest < ActiveSupport::TestCase
```
Comment (gagan93jtg) [inline test/models/task_comment_test.rb]: add rspec right now or don't commit test files for now. Delete them after autogeneration. The default picked library is minitest and we don't use it.

**Review by gagan93jtg — CHANGES_REQUESTED** [review]

File: db/migrate/20230821083754_add_not_null_constraints_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class AddNotNullConstraintsToUsers < ActiveRecord::Migration[7.0]
+  def change
+    change_column_null :users, :email, false
```
Comment (neeraj33negi) [inline db/migrate/20230821083754_add_not_null_constraints_to_users.rb:3]: what does this do? just curious

File: db/migrate/20230821083659_create_tasks_task_categories.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class CreateTasksTaskCategories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :tasks_task_categories do |t|
+      t.references :tasks, null: false, foreign_key: true
```
Comment (neeraj33negi) [inline db/migrate/20230821083659_create_tasks_task_categories.rb]: is this correct? should this be `t.references :task ...` ?

File: db/migrate/20230821083532_create_tasks.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class CreateTasks < ActiveRecord::Migration[7.0]
+  def change
+    create_table :tasks do |t|
+      t.string :title
+      t.string :description
+      t.date :due_date
+      t.integer :priority
+      t.integer :status
+      t.references :assignee_user, null: false, foreign_key: true
+      t.references :assigned_user, null: false, foreign_key: true
```
Comment (neeraj33negi) [inline db/migrate/20230821083532_create_tasks.rb]: did this migration run?

File: db/migrate/20230821083532_create_tasks.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class CreateTasks < ActiveRecord::Migration[7.0]
+  def change
+    create_table :tasks do |t|
+      t.string :title
+      t.string :description
+      t.date :due_date
+      t.integer :priority
+      t.integer :status
```
Comment (neeraj33negi) [inline db/migrate/20230821083532_create_tasks.rb:8]: add default/initial values

File: db/migrate/20230821083518_create_users.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+class CreateUsers < ActiveRecord::Migration[7.0]
+  def change
+    create_table :users do |t|
+      t.string :email
```
Comment (neeraj33negi) [inline db/migrate/20230821083518_create_users.rb:4]: unique index

File: config/database.yml [hunk 8]
```diff
@@ -5,21 +5,24 @@
...
 # Do not set this db to the same as development or production.
 test:
   <<: *default
-  database: db/test.sqlite3
+  database: TaskManagement_test
 
 production:
   <<: *default
-  database: db/production.sqlite3
+  database: TaskManagement_production
```
Comment (neeraj33negi) [inline config/database.yml]: snake case

File: app/models/task_comment.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class TaskComment < ApplicationRecord
+  belongs_to :user
+  belongs_to :task
+end
```
Comment (neeraj33negi) [inline app/models/task_comment.rb]: I dont see has_many associations anywhere

File: db/migrate/20230821083754_add_not_null_constraints_to_users.rb (hunk 3 above)
Comment (lakshay42jain) [reply db/migrate/20230821083754_add_not_null_constraints_to_users.rb:3]: email should not be null

File: app/models/task.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  has_many :task_task_categories
+  has_many :task_categories, through: :task_task_categories
+
+  validates :title, presence: true
+  validates :description, presence: true
+  validates :due_date, presence: true
+  validates :priority, presence: true
+  validates :status, presence: true
+  validates :assignee_user_id, presence: true
+  validates :assigner_user_id, presence: true
```
Comment (neeraj33negi) [inline app/models/task.rb]: these could be one liner, pls check how

File: app/models/task.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+class Task < ApplicationRecord
+  enum status: {pending: 1, in_progress: 2, completed: 3}
+  belongs_to :user, foreign_key: 'assignee_user_id'
+  belongs_to :user, foreign_key: 'assigner_user_id'
+  has_many :task_comments
+  has_many :task_task_categories
```
Comment (neeraj33negi) [inline app/models/task.rb]: add `dependent`

File: app/models/task.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
+class Task < ApplicationRecord
+  enum status: {pending: 1, in_progress: 2, completed: 3}
```
Comment (neeraj33negi) [inline app/models/task.rb]: space after and before `{` and `}`
add empty line after enum definition

File: app/models/task_category.rb [hunk 13]
```diff
@@ -0,0 +1,4 @@
+class TaskCategory < ApplicationRecord
+  has_many :task_task_categories
```
Comment (neeraj33negi) [inline app/models/task_category.rb]: dependent

File: app/models/task_category.rb [hunk 14]
```diff
@@ -0,0 +1,4 @@
+class TaskCategory < ApplicationRecord
```
Comment (neeraj33negi) [inline app/models/task_category.rb:1]: check if there should be validations here

File: app/models/task.rb (hunk 10 above)
Comment (neeraj33negi) [reply app/models/task.rb]: also check what other validations should be there

File: app/models/task.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  belongs_to :user, foreign_key: 'assignee_user_id'
+  belongs_to :user, foreign_key: 'assigner_user_id'
```
Comment (neeraj33negi) [inline app/models/task.rb]: `belongs_to :user` twice? this is ambiguous, check how this could be improved

File: app/models/user.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class User < ApplicationRecord
+  enum role: {admin: 0, normal: 1}
+  has_many :tasks, foreign_key: 'assignee_user_id'
+  has_many :task_comments
+
+  validates :email, presence: true, uniqueness: true
+  validates :phone_number, presence: true
+  validates :name, presence: true
+  validates :password, presence: true
+  validates :active, inclusion: { in: [true, false] }
```
Comment (neeraj33negi) [inline app/models/user.rb]: is this needed?

Comment (neeraj33negi) [reply app/models/user.rb]: check all the validations that can be applied here

**Review by neeraj33negi — CHANGES_REQUESTED** [review]

File: app/models/task_category.rb (hunk 14 above)
Comment (neeraj33negi) [reply app/models/task_category.rb:1]: read about `accepts_nested_attributes` and add validation for task_categories

File: app/models/task.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  private def status_validate
+    if completed?
+      self.errors[:base] << "New tasks cannot be created with status completed"
+      throw(abort)
+    end
+  end
+
+  private def assign_task_category
+    self.task_categories << TaskCategory.find(1)
+  end
```
Comment (neeraj33negi) [inline app/models/task.rb]: why is this here? also why are we hardcoding a db id in the code?

File: app/models/task_task_category.rb [hunk 18]
```diff
@@ -0,0 +1,4 @@
+class TaskTaskCategory < ApplicationRecord
```
Comment (neeraj33negi) [inline app/models/task_task_category.rb:1]: add a uniqness validation, there shud be just on entry in join table for a specific pair of task and task_category, read https://boringrails.com/tips/rails-unique-scope

File: app/models/task.rb [hunk 19]
```diff
@@ -0,0 +1,48 @@
+class Task < ApplicationRecord
+  before_create :assign_task_category
+
+  enum status: { pending: 1, in_progress: 2, completed: 3 }
+
+  belongs_to :assignee_user, class_name: 'User'
```
Comment (gagan93jtg) [inline app/models/task.rb]: `assignee` and `assigner` are ok

File: app/models/task.rb [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  belongs_to :assigner_user, class_name: 'User'
+
+  has_many :task_comments, dependent: :destroy
```
Comment (gagan93jtg) [inline app/models/task.rb]: can we simply call them "comments"?

File: app/models/task.rb [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+  has_many :task_task_categories, dependent: :destroy
+  has_many :task_categories, through: :task_task_categories
```
Comment (gagan93jtg) [inline app/models/task.rb]: and `categories`?

File: app/models/task.rb [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+
+  validate :admin_validate
+  validate :status_validate, on: :create
+  validate :due_date_format_and_range
+  validates_presence_of :title, :description, :due_date, :priority, :status
+
+  private def admin_validate
+    unless assigner_user.admin?
+      self.errors[:base] << 'Only admin can create the task'
+      raise StandardError, 'Only admin can create the task'
```
Comment (gagan93jtg) [inline app/models/task.rb]: why are we raising errors here? and why are we repeating strings in both lines?

File: app/models/user.rb [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+  enum role: { admin: 0, normal: 1 }
+
+  has_many :tasks, foreign_key: 'assignee_user_id', dependent: :destroy
+  has_many :task_comments, dependent: :destroy
+
+  validates_presence_of :name, :password
+  validates :phone_number, presence: true, length: { is: 10 }
+  validates :email, presence: true, uniqueness: true, format: { with: /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i }
+
+  private
```
Comment (gagan93jtg) [inline app/models/user.rb]: avoid private block. Use `private` with specific method if u want to declare as "private".

**Review by gagan93jtg — CHANGES_REQUESTED** [review]

**Review by PramanSatya — APPROVED** [review]

**Pull request merged**